### PR TITLE
Refactor the client

### PR DESCRIPTION
**Pull request by easternbloc**
Moved the codes about into the correct locations.

We now have three modules to describe the platform:
* Dashboard
* Module
* Datasource

and two views for the data:

* Table
* Delta

*Note* This is full of breaking API changes. The shape of the data has understandably changed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,17 +4,57 @@
...
+```
+
+## Module
+
+```
+
+var Module = require('performaceplatform-client.js').Module,
+    module = new Module(moduleConfig);
+
+module.resolve().then(function (dashboard) {
```
Comment (tombooth) [inline README.md]: Shouldn't the `dashboard` argument by `module`?

Comment (easternbloc) [inline README.md]: That might be a copy paste error there... :)

Will fix

**Comment by jonnywyatt** [comment]
Could we get some tests comparing their output against test/fixtures/dashboard-processed.json?
Then we'll have some coverage against both sides of that contract.

File: README.md [hunk 2]
```diff
@@ -4,17 +4,57 @@
...
+## Dashboard
+
+```
 
-new Dashboard().getConfig('dashboard-slug')
-.then(function (dashboardConfig) {
-  var dashboard = JSON.parse(dashboardConfig);
-  // Do something with the dashboard data
-  // ...
+var Dashboard = require('performaceplatform-client.js').Dashboard,
```
Comment (jonnywyatt) [inline README.md]: typo in the package name ("performaceplatform"), is this working?

File: lib/Dashboard.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+Dashboard.prototype.getConfig = function () {
+  var options = _.extend(this.options, {
+    url: config.stagecraft + 'public/dashboards?slug=' + this.options.slug
+  });
+
+  return rp(options);
+};
+
+Dashboard.prototype.resolve = function () {
```
Comment (jonnywyatt) [inline lib/Dashboard.js]: Could the naming of getConfig and resolve better indicate their relationship? 
getConfig gets a 'shallow' dashboard config without modules
resolve gets that plus modules
the names just seem unrelated, could make the API more intuitive?

Comment (easternbloc) [inline lib/Dashboard.js]: I wanted a standard name for "do everything" that could be used through the dashboard and module, modules. I was also hoping that as a user you didn't care too much about what's going on under the hood. I'd be interested to see what other peoples thoughts are on this? I'm open to suggestions around this? Do you have any thoughts on what it could be renamed too?

File: README.md (hunk 2 above)
Comment (easternbloc) [inline README.md]: It's totally tested honest! Will update.

**Comment by easternbloc** [comment]
@jonnywyatt yep. Will do. I forgot about test/fixtures/dashboard-processed.json it needs updating as the shape of the data has changed.

File: lib/Dashboard.js (hunk 3 above)
Comment (jonnywyatt) [inline lib/Dashboard.js]: how about
getConfig / getConfigAndModules

Comment (tombooth) [inline lib/Dashboard.js]: Agreed i don't know if resolve() could gain a shallow and a deep version ie:

`dashboard.resolve(true)` would recurse down. That doesn't read the best but maybe it can be a constant or string or something.

**Comment by tombooth** [comment]
Once last request, because it is rather painful. The applications should inject the configuration (stagecraft/backdrop URLs) not the client. At the moment this means we can't easily have preview/staging/production deployments of applications pointing at the correct environment.

**Comment by tombooth** [comment]
Just waiting on the config change, then I'll happily merge!

**Comment by tombooth** [comment]
Version bump?

**Comment by easternbloc** [comment]
@tombooth done.

**Comment by jonnywyatt** [comment]
Could test/fixtures/dashboard-processed.json be updated and get some tests against it?

**Comment by tombooth** [comment]
I don't think dashboard-processed.json is really representative of how the client works now and should probably be reviewed.

**Comment by jonnywyatt** [comment]
Isn't it the JSON that's output from the resolve() method (formerly getDashboardMetrics)?

**Comment by easternbloc** [comment]
We now have views on the data. So resolve just returns the data but it isn't formatted.

**Comment by easternbloc** [comment]
There are a number of different entry points to the data so having one file for this now doesn't make sense.

**Comment by jonnywyatt** [comment]
Cool - could we at least agree on the principle that both the client and users of the client could run tests against the same contract / fixture(s)?
